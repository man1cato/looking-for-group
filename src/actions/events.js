import axios from 'axios';
import moment from 'moment';
import _ from 'lodash';

const apiKey= 'keyzG8AODPdzdkhjG';
const baseUrl = 'https://api.airtable.com/v0/appOY7Pr6zpzhQs6l';

//GET_EVENTS
export const getEvents = (events) => ({
    type: 'GET_EVENTS',
    events
});

export const startGetEvents = (usersGroups, userAvailabilities) => {
    return async (dispatch) => {
        try {
            let events = [];
            for (let group of usersGroups) {                                    //FOR EACH OF THE USER'S GROUPS...
                const groupRecordId = group.id; 
                for (let availabilityRecordId of userAvailabilities) {                    
                    const filter = `AND({Group Record ID}="${groupRecordId}",{Availability Record ID}="${availabilityRecordId}")`;
                    const eventResponse = await axios.get(`${baseUrl}/Events?filterByFormula=${filter}&api_key=${apiKey}`); 
                    const event = eventResponse.data.records[0];
                    if (event) {
                        events.push({
                            id: event.id, 
                            interest: group.interest,
                            area: group.area,
                            availability: event.fields.Availability[0],
                            occurrence: event.fields.Occurrence,
                            startDatetime: event.fields['Start Date & Time'],
                            date: moment(event.fields['Start Date & Time']).format('MMM Do'),
                            startTime: moment(event.fields['Start Date & Time']).format('h:mm a')
                        });
                    }
                }
            }
            console.log('events:', events);
            const sortedEvents = _.orderBy(events, ['startDatetime'], ['asc']);
            dispatch(getEvents((sortedEvents)));
        } catch (e) {
            throw new Error('Failed to retrieve groups in startGetEvents');
        }
    };
};


// //UPDATE_EVENTS
// export const updateEvents = (groups) => ({
//     type: 'UPDATE_GROUPS',
//     groups
// });

// export const startUpdateGroups = (areaRecordId, userInterests) => {
//     return async (dispatch) => {
//         try {
//             let groups = [];
//             for (let interestRecordId of userInterests) {
//                 const filter = `AND({Interest Record ID}="${interestRecordId}",{Area Record ID}="${areaRecordId}")`;
//                 const response = await axios.get(`${baseUrl}/Groups?filterByFormula=${filter}&api_key=${apiKey}`);
//                 const group = response.data.records[0];
//                 groups.push({
//                     id: group.id, 
//                     area: group.fields["Area Text"],
//                     interest: group.fields["Interest Name"],
//                     availability: group.fields["Group Availibility"]
//                 });
//             }
//             const sortedGroups = _.orderBy(groups, ['interest'], ['asc']);
//             console.log('from startUpdateGroups:',sortedGroups);
//             dispatch(updateGroups(sortedGroups));
//         } catch (e) {
//             throw new Error('Update groups failed at startUpdateGroups');
//         }
        
//     };
// };