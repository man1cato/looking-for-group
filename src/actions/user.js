import axios from 'axios';

const baseUrl = 'https://api.airtable.com/v0/appOY7Pr6zpzhQs6l';
const apiKey= 'keyzG8AODPdzdkhjG';


//SET_USER
export const setUser = (user = {}) => ({
    type: 'SET_USER',
    user
});

export const startSetUser = ({uid, email}) => {
    const filterFormula = `{Email}="${email}"`;
    return async (dispatch) => {
        try {
            //SEARCH AIRTABLE FOR USER'S EMAIL
            const response = await axios.get(`${baseUrl}/Users?filterByFormula=${filterFormula}&api_key=${apiKey}`);
            //IF EMAIL FOUND, GATHER PROFILE INFO
            if (response.data.records.length > 0) {
                const userRecord = response.data.records[0];
                const interest1 = userRecord.fields['#1 Interest'][0] ? userRecord.fields['#1 Interest'][0] : [];
                const interest2 = userRecord.fields['#2 Interest'][0] ? userRecord.fields['#2 Interest'][0] : [];
                const interest3 = userRecord.fields['#3 Interest'][0] ? userRecord.fields['#3 Interest'][0] : [];
                const additionalInterests = userRecord.fields["Add'l Interests"] ? userRecord.fields["Add'l Interests"] : [];
                const allInterests = [interest1, interest2, interest3, ...additionalInterests];
                const user = {
                    recordId: userRecord.id,
                    firstName: userRecord.fields['First Name'],
                    lastName: userRecord.fields['Last Name'],
                    email,
                    postalCode: userRecord.fields['Postal Code'],
                    birthYear: userRecord.fields['Birth Year'],
                    interest1,
                    interest2,
                    interest3,
                    additionalInterests,
                    allInterests,
                    availability: userRecord.fields.Availability,
                    area: userRecord.fields.Area[0],
                    groups: userRecord.fields.Groups
                };
                //IF FIREBASE ID NOT PRESENT IN RECORD, ADD TO RECORD
                if(!response.data.records[0].fields['Firebase ID']) {
                    axios.patch(`${baseUrl}/Users/${user.recordId}?api_key=${apiKey}`, {"fields": {"Firebase ID": uid} } );
                }
                dispatch(setUser(user));
                return user;
            } else {
                const postResponse = await axios.post(`${baseUrl}/Users?api_key=${apiKey}`, {"fields": {"Email": email, "Firebase ID": uid}});
                const user = {
                    recordId: postResponse.data.id,
                    email
                };
                dispatch(setUser(user));
                return user;
            }
        } catch (e) {
            throw new Error('Call to Airtable Users table failed');
        }
    };
};

//UPDATE_USER
export const updateUser = (user) => ({
    type: 'UPDATE_USER',
    user
});

export const startUpdateUser = (user) => {
    const fields = {
        "First Name": user.firstName,
        "Last Name": user.lastName,
        "Postal Code": user.postalCode,
        "Birth Year": Number(user.birthYear),
        "#1 Interest": [user.interest1],
        "#2 Interest": [user.interest2],
        "#3 Interest": [user.interest3],
        "Add'l Interests": user.additionalInterests,
        "Availability": user.availability
    };
    return (dispatch) => {
        if (user.recordId) {
            axios.patch(`${baseUrl}/Users/${user.recordId}?api_key=${apiKey}`, {"fields": fields});
        } else {
            axios.post(`${baseUrl}/Users?api_key=${apiKey}`, {"fields": {"Email": user.email, "Firebase ID": user.uid, ...fields}});
        }
        dispatch(updateUser(user));
    };
};


// //UPDATE_USER_AREA  
// export const updateUserArea = (area) => ({
//     type: 'UPDATE_USER_AREA',
//     area
// });

// export const startUpdateUserArea = (user, placeDetails) => {
//     // const userRecordId = user.recordId;
//     // const userInterests = user.allInterests;
//     return (dispatch) => {
//         updateAirtable(user, placeDetails);
//         dispatch(updateUser(user));
//     };
//     // return async (dispatch) => {
//     //     try {
//     //         const areaRecordId = await geolocateUser(userRecordId, placeDetails);
//     //         const usersGroups = await addUserToGroups(userRecordId, userInterests, areaRecordId);
//     //         // for (let groupRecordId of usersGroups) {                                        //3. UPDATE GROUP AVAILABILITY
//     //         //     const group = await axios.get(`${baseUrl}/Groups/${groupRecordId}?api_key=${apiKey}`);
//     //         //     await updateGroupAvailability(group, userAvailability);                                 
//     //         // }
//     //         dispatch(updateUser({...user, area: areaRecordId, groups: usersGroups}));
//     //     } catch (e) {
//     //         throw new Error('Error at addUserToGroups within startUpdateUserArea');
//     //     }
//     // };
// };
