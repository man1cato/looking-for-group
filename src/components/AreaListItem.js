import React from 'react';
import {Link} from 'react-router-dom';

const AreaListItem = ({id, name}) => (
    <Link className="list-item" to={`/areas/${id}`}>                             
        <div>
            <h3 className="list-item__title">{name}</h3>
        </div>
    </Link>
);


export default AreaListItem;