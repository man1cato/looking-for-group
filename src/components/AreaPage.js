import React from 'react';
import {connect} from 'react-redux';
import getAreas from '../selectors/areas';

export const AreaPage = (props) => (
    <div>
        <h1>Areas</h1>
        <div>
            Put areas here
        </div>
    </div>
);

// const mapStateToProps = (state) => {
//     const view = 'Alpha Test List';
//     return {
//         areas: getAreas(view)
//     };
// };

// export default connect(mapStateToProps)(AreaPage);
