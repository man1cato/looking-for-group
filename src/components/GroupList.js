import React from 'react';
import {connect} from 'react-redux';
import GroupListItem from './GroupListItem';


export const GroupList = (props) => (
    <div>
        <div className="page-header">
            <div className="content-container">
                <h2 className="page-header__title">My Groups</h2>
                <span>{props.groups[0].area}</span>
            </div>
        </div>
        <div className="content-container">
            <div className="list-body">
                {
                    props.groups.length === 0 ? (
                        <div className="list-item list-item--message">Please update your profile to see available groups in your area.</div>
                    ) : (
                        props.groups.map((group) => <GroupListItem key={group.id} {...group} /> )    
                    )
                }
            </div>
        </div>
    </div>
);


// const mapStateToProps = (state) =>  state.areas;

const mapStateToProps = (state) => ({ groups: state.user.groups });

export default connect(mapStateToProps)(GroupList);
