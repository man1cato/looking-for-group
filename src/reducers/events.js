//EVENTS REDUCER
const eventsReducer = (state = [], action) => {
    switch (action.type) {
        case 'GET_EVENTS':
            return action.events;
        case 'UPDATE_EVENTS':
            return action.events;
        default: 
            return state;
    }
};

export default eventsReducer;