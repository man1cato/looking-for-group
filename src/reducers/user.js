//USER REDUCER

const userReducerDefaultState = {                                                  //DEFINE USER OBJECT
    recordId: '',
    firstName: '',
    lastName: '',
    email: '',
    postalCode: '',
    birthYear: 1990,
    interest1: '',
    interest2: '',
    interest3: '',
    additionalInterests: [],
    allInterests: [],
    availability: [],
    area: '',
    groups: []
};

const userReducer = (state = userReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_USER':
            return action.user;
        default:
            return state;
    }
};

export default userReducer;