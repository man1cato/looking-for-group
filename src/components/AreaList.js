import React from 'react';
import {connect} from 'react-redux';
import AreaListItem from './AreaListItem';


export const AreaList = (props) => (
    <div>
        <div className="page-header">
            <div className="content-container">
                <h2 className="page-header__title">Areas</h2>
                <span>Select the area in which you live.</span>
            </div>
        </div>
        <div className="content-container">
            <div className="list-body">
                {
                    props.areas.length === 0 ? (
                        <div className="list-item list-item--message">No areas</div>
                    ) : (
                        props.areas.map((area) => <AreaListItem key={area.id} {...area} /> )    
                    )
                }
            </div>
        </div>
    </div>
);


// const mapStateToProps = (state) =>  state.areas;

const mapStateToProps = (state) => ({ areas: state.areas });

export default connect(mapStateToProps)(AreaList);
