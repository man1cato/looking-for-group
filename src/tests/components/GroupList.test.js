import React from 'react';
import {shallow} from 'enzyme';
import {GroupList} from '../../components/GroupList';
import user from '../fixtures/user';
import group from '../fixtures/group';
import events from '../fixtures/events';


test('should render GroupList message for a user without groups', () => {
    const wrapper = shallow(<GroupList groups={[]} events={[]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render GroupList for a user with groups', () => {
    const wrapper = shallow(<GroupList groups={user.groups} area={user.area} events={events}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render GroupList message for a user with groups but no matching events', () => {
    const wrapper = shallow(<GroupList groups={user.groups} area={user.area} events={[]}/>);
    expect(wrapper).toMatchSnapshot();
});