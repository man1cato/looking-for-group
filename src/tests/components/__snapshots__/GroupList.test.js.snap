// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render GroupList for a user with groups 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
        My Local Interest Groups
      </h2>
      <div>
        Groups will only appear for those interests that have a sufficient number of people in the area who share that interest.
      </div>
    </div>
  </div>
  <div
    className="content-container"
  >
    <div
      className="list-body"
    >
      <GroupListItem
        area="\\"Miami-Fort Lauderdale-West Palm Beach, FL\\""
        availability={
          Array [
            "rectsqF1uDWuxJGWQ",
            "rec4hxrJGMNk8I5RP",
            "recPljFv5l1JzHCbP",
            "recL5BIcv756EQ7oC",
            "recDniSFi8uMfboQx",
            "recmiujOyHwlG42hH",
            "recDmj0waZL3td5fS",
          ]
        }
        eventCount={7}
        events={
          Array [
            "recshxa9OHYHAAoVt",
            "rec22jQuu4zqD8gs3",
            "recLfH8xljwjj3Evz",
            "rec2df6Ju5CxATj45",
            "rechXOUEJWTQHGwXg",
            "rec1bJ6FzRCLCHV89",
            "recmeOzk6VxhIsu9h",
          ]
        }
        id="recTX8o0BxS4iaPQr"
        interest="Video Games"
        key="recTX8o0BxS4iaPQr"
        timezoneId="America/New_York"
        userCount={5}
        userIds={
          Array [
            "recgCmMoJzRg7XMlU",
            "reco3FETF7HxNIWfv",
            "recZZaSE19c9cGypA",
            "recbkFiE7d9OM1elO",
            "recb4V06Qm8MvMbJ1",
          ]
        }
      />
    </div>
  </div>
</div>
`;

exports[`should render GroupList message for a user with groups but no matching events 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
        My Local Interest Groups
      </h2>
      <div>
        Groups will only appear for those interests that have a sufficient number of people in the area who share that interest.
      </div>
    </div>
  </div>
  <div
    className="content-container"
  >
    <div
      className="list-body"
    >
      <div
        className="list-item list-item--message"
      >
        <div>
          Apologies, but it seems there are insufficient users in your area who match your interests.
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render GroupList message for a user without groups 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
        My Local Interest Groups
      </h2>
      <div>
        Groups will only appear for those interests that have a sufficient number of people in the area who share that interest.
      </div>
    </div>
  </div>
  <div
    className="content-container"
  >
    <div
      className="list-body"
    >
      <div
        className="list-item list-item--message"
      >
        <div>
          Please update your 
          <span>
            <Link
              replace={false}
              to="/profile"
            >
              profile
            </Link>
          </span>
           to see available groups in your area. If you've already done so, please give this a moment to load.
        </div>
      </div>
    </div>
  </div>
</div>
`;
